src/OVAL/probes/unix/linux/rpmverifyfile_probe.c

```diff
@@ -237,92 +321,27 @@ static int rpmverify_collect(probe_ctx *ctx,
 			oscap_streq(res.epoch, "(none)") ? "0" : res.epoch,
 			res.version, res.release, res.arch);
 
-		/*
-		 * Inspect package files & directories
-		 */
-		for (i = 0; i < 2; ++i) {
-			fi = rpmfiNew(g_rpm->rpmts, pkgh, tag[i], 1);
-
-		  while (rpmfiNext(fi) != -1) {
-				current_file = rpmfiFN(fi);
-				current_file_realpath = oscap_realpath(current_file, NULL);
-		    res.fflags = rpmfiFFlags(fi);
-		    res.oflags = omit;
-
-		    if (((res.fflags & RPMFILE_CONFIG) && (flags & RPMVERIFY_SKIP_CONFIG)) ||
-					((res.fflags & RPMFILE_GHOST)  && (flags & RPMVERIFY_SKIP_GHOST))) {
-					free(current_file_realpath);
-					continue;
-				}
-
-		    switch(file_op) {
-		    case OVAL_OPERATION_EQUALS:
-					if (strcmp(current_file, file) != 0 &&
-							current_file_realpath && file_realpath &&
-							strcmp(current_file_realpath, file_realpath) != 0) {
-						free(current_file_realpath);
-						continue;
-					}
-					res.file = oscap_strdup(file);
-		      break;
-		    case OVAL_OPERATION_NOT_EQUAL:
-					if (strcmp(current_file, file) == 0 ||
-							(current_file_realpath && file_realpath &&
-							strcmp(current_file_realpath, file_realpath) == 0)) {
-						free(current_file_realpath);
-						continue;
-					}
-					res.file = current_file_realpath ? current_file_realpath : strdup(current_file);
-		      break;
-		    case OVAL_OPERATION_PATTERN_MATCH:
-					ret = pcre_exec(re, NULL, current_file, strlen(current_file), 0, 0, NULL, 0);
-
-		      switch(ret) {
-		      case 0: /* match */
-						res.file = strdup(current_file);
-			break;
-		      case -1:
-			/* mismatch */
-						free(current_file_realpath);
-			continue;
-		      default:
-			dE("pcre_exec() failed!");
+		if (rpmverify_collect_package_files_or_directories(g_rpm, ctx, pkgh, file, file_op, RPMTAG_BASENAMES, &res, flags) != 0) {
 			ret = -1;
-						free(current_file_realpath);
+		}
+		if (rpmverify_collect_package_files_or_directories(g_rpm, ctx, pkgh, file, file_op, RPMTAG_DIRNAMES, &res, flags) != 0) {
+			ret = -1;
+		}
```

Comment:
Does it make sense to call `rpmverify_collect_package_files_or_directories` a second time with `RPMTAG_DIRNAMES` if the call with `RPMTAG_BASENAMES` resulted in error (`ret = -1`)?

In the original code with the `for` loop, whenever there was an error it would go to `cleanup`, exiting the loop and stop collecting package files or directories.